GaugePanel: Setting the neutral-point of a gauge

sfranzis opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
The neutral or zero point of a gauge is at the left of the panel. When I set the range from e.g. Min=-5 to Max = 5 then the display of negative values looks not natural.
I would like the neutral-point to be configurable in the standard field options (where we have Min, Max, ...).

Example use case: a solar battery inverter, which shows positive values when charging and negative values when discharging.

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana/discussions/38273

Usage: If min and max of a gauge panel are additive inverse numbers, e.g. min=-500, max=500 then the neutral point of the gauge is put at 12 o'clock. Negative values are plotted to the left side, positive values to the right side.

sfranzis committed:
GraphPanel: Setting the neutral-point of a gauge

If min and max of a gauge panel are opposite numbers, e.g. min=-500, max=500 then the neutral point of the gauge is put at 12 o'clock.
Negative values are plotted to the left side, positive values to the right side.
See Gitlab discussions  #38273

sfranzis committed:
GraphPanel: Setting the neutral-point of a gauge

remove unneded const angularModules

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=53989) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=53989) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=53989) it.</sub>

sfranzis committed:
Update gauge docs

tclausing commented:
I'm happy to see this feature. Though, limiting to inverse min/max scenario seems arbitrary. Why not expose a setting for "fill from zero" regardless of the angle of the zero point?

mellieA commented:
@sfranzis thank you for this contribution! I've moved this to the to the User Essentials dev project for review and comments

mellieA commented:
Approvers, here are some images copied over from the feature discussion that this PR resolves: 

![image](https://user-images.githubusercontent.com/18175287/189451897-2d0967ce-274f-41aa-8b8a-e506763a7e5b.png)

sfranzis commented:
> I'm happy to see this feature. Though, limiting to inverse min/max scenario seems arbitrary. Why not expose a setting for "fill from zero" regardless of the angle of the zero point?
> 
> Perhaps this should even be default behavior of a gauge. I can't think of an example where painting from a negative min to a positive value would be desired since a gauge is meant to show relative magnitudes, which implies painting from zero to represent magnitude correctly.

An additional setting would require changes on the data model and more. For this I'm not yet deep enough in the code. This is my first change.
I agree with the idea making this to the default behavior. But also in this case you need to set min and max values to have a fixed scale.

gguillotte-grafana commented on docs/sources/visualizations/gauge-panel.md:
```diff
@@ -53,3 +53,10 @@ Adjust the sizes of the gauge text.
+
+Adjust various options.
+
+- **Min/Max -** If the values of min and max are additive inverse numbers (e.g. min=-500, max=500), then the neutral point of the gauge is put at 12 o'clock. Negative values are plotted to the left side, positive values to the right side.
```
Aiming for clarity with plainer language here, while understanding the limitations in the comments:

```suggestion
- **Min/Max -** Define a range to gauge, with the neutral point plotted at the center of the gauge. Values higher or lower than the neutral point plotted to the left or right of center, respectfully. The values of min and max must be opposite numbers; for instance, if the min is -500, the max must be 500.
```

gguillotte-grafana commented on public/vendor/flot/jquery.flot.gauge.js:
```diff
@@ -344,20 +343,48 @@
 
+
+        /**
+         * if gauge min is negative and has the absolute value of min and max are the same
```
```suggestion
         * if gauge min is negative and the absolute value of min and max are the same
```

gguillotte-grafana commented on docs/sources/visualizations/gauge-panel.md:
```diff
@@ -53,3 +53,10 @@ Adjust the sizes of the gauge text.
 - **Title -** Enter a numeric value for the gauge title size.
 - **Value -** Enter a numeric value for the gauge value size.
+
+## Standard options
+
+Adjust various options.
```
With only one setting listed, "various options" doesn't work well here. If I understand the comments, this is changing a default behavior of gauges, so would "Adjust default gauge behaviors" work instead?

sfranzis replied:
Hi @gguillotte-grafana thanks for your review.

gguillotte-grafana reviewed (commented):
There are some phrasing changes that could make the docs addition clearer and require less domain knowledge to apply.

JoaoSilvaGrafana commented on public/vendor/flot/jquery.flot.gauge.js:
```diff
@@ -344,20 +343,48 @@
         }
 
+
+        /**
+         * if gauge min is negative and has the absolute value of min and max are the same
+         * then calculate a1 and a2 in a way, that zero is at 12 o'clock
+         * 
+         * @method calculateAnglesForGauge
+         * @param {Object} gaugeOptionsi the options of the gauge
+         * @returns {Object}
+         */
+        function calculateAnglesForGauge(gaugeOptionsi, layout, data) {
+            let angles = {};
+            var setNorth = (gaugeOptionsi.gauge.min + gaugeOptionsi.gauge.max) == 0.0
+            
+            angles.a1 = gaugeOptionsi.gauge.startAngle;
+            angles.a2 = calculateAngle(gaugeOptionsi, layout, data);
+
+            if(setNorth && angles.a2<=1.5) {
+                angles.a1 = angles.a2;
+                angles.a2 = 1.5;
+            }
+            if(setNorth && angles.a2>1.5) {
+                angles.a1 = 1.5;
+            }
+            
+            return angles;
+        }
```
Hey @sfranzis thank you for contributing to Grafana! 🎉 

I have to agree with the sentence "limiting to inverse min/max scenario seems arbitrary", I think it would end up being almost a "hidden feature" that if you set min and max to the same absolute value you get this extra feature on the Gauges that have negative numbers.

I think ideally this should always happen if the minimum of the gauge is < 0, and then we should be able to see some kind of indication of where 0 is as well, maybe with a dotted line or with the actual number 0 next to that point.
So I feel like in order for us to merge this it would need a little bit more work. This is my suggestion for what this function should look like:
```
        /**
         * Calcualte the angles for the gauge, depending on if there are
         * negative numbers or not.
         * 
         * @method calculateAnglesForGauge
         * @param {Object} gaugeOptionsi the options of the gauge
         * @param  {Number} data the value of the gauge
         * @returns {Object}
         */
        function calculateAnglesForGauge(gaugeOptionsi, layout, data) {
            let angles = {};
            var hasNegative = gaugeOptionsi.gauge.min < 0.0
            var isNegative = data < 0.0
            
            if (hasNegative) {
                if (isNegative) {
                    angles.a1 = calculateAngle(gaugeOptionsi, layout, data);
                    angles.a2 = calculateAngle(gaugeOptionsi, layout, 0.0);
                } else {
                    angles.a1 = calculateAngle(gaugeOptionsi, layout, 0.0);
                    angles.a2 = calculateAngle(gaugeOptionsi, layout, data);
                }
            } else {
                angles.a1 = gaugeOptionsi.gauge.startAngle;
                angles.a2 = calculateAngle(gaugeOptionsi, layout, data);
            }
            
            return angles;
        }
```

What do you think? Then the only thing left would be to find a way to represent where the 0 is when the min is negative. Would you like to give that a go?

sfranzis replied:
Hey @JoaoSilvaGrafana thanks for your suggestion. I'm fine with it. 👍

sfranzis commented:
Working on it.
1. Regarding the modified logic, (when minimum < 0) it works is general, but is a bit tricky when no values for min/max are set. In this case the values vor min/max are calculated. Any idea how to check if no value is set in the GUI?
2. Zero marker - I implemented this:
<img width="550" alt="image" src="https://user-images.githubusercontent.com/12957935/190902105-b8d81f48-e20d-427e-bc5f-f20e4d68b821.png">

LarsStegman commented:
Thanks for the pull request. Just my 2 cents: I would prefer to be able to override the neutral point to set it to any arbitrary value. 0 being the default is fine with me.

JoaoSilvaGrafana commented:
> Working on it.
> 
> 1. Regarding the modified logic, (when minimum < 0) it works is general, but is a bit tricky when no values for min/max are set. In this case the values vor min/max are calculated. Any idea how to check if no value is set in the GUI?
> 2. Zero marker - I implemented this:
> 
> <img alt="image" width="550" src="https://user-images.githubusercontent.com/12957935/190902105-b8d81f48-e20d-427e-bc5f-f20e4d68b821.png">

Hey @sfranzis sorry for the delay, been on holiday 😄 
1. I'm not sure I understand your point, could you give me an example of where it doesn't work? I tried it with various values and no min/max and it behaves correctly:
<img width="1457" alt="Screenshot 2022-09-30 at 16 20 10" src="https://user-images.githubusercontent.com/100691367/193303034-27d4f74c-7186-47cc-983f-893c39c02261.png">
2. I think we might need to put this through a UX check, could you push your changes and I can ask some designers here and see what they think? 🙂 

Thanks for working on this!

JoaoSilvaGrafana commented:
> Thanks for the pull request. Just my 2 cents: I would prefer to be able to override the neutral point to set it to any arbitrary value. The current implementation would be fine for the default, i.e. the neutral point is always at min, unless the user overrides it.

Hey @LarsStegman , just trying to understand your point. Could you give an example of what you expected to happen with some values? By neutral point do you mean the point in the middle of the Gauge?

LarsStegman commented:
> > Thanks for the pull request. Just my 2 cents: I would prefer to be able to override the neutral point to set it to any arbitrary value. The current implementation would be fine for the default, i.e. the neutral point is always at min, unless the user overrides it.
> 
> Hey @LarsStegman , just trying to understand your point. Could you give an example of what you expected to happen with some values? By neutral point do you mean the point in the middle of the Gauge?

Hi, I would like the "zero point" settable. For example:
* the current value is 3
* The minimum value is -5
* The maximum value is 5
* the neutral point is manually set to -1

The gauge would then be drawn between -1 and 3

of course "zero point" is not accurate then, so I would name it "neutral point" or something like that.

sfranzis committed:
Setting the neutral-point of a gauge. make it default behaviour and add zero marker

sfranzis committed:
Setting the neutral-point of a gauge. make it default behaviour and add zero marker

sfranzis committed:
Remove hint to set min/max as the neutral point is now set automatically

sfranzis commented:
> 2. I think we might need to put this through a UX check, could you push your changes and I can ask some designers here and see what they think? 🙂
> Thanks for working on this!

Sorry for the delay, to much real life things 🙈
Implemented and pushed my changes now. Please see and let me know what you think. :-)

sfranzis committed:
Merge branch 'grafana:main' into #38273

sfranzis committed:
Merge branch 'grafana:main' into #38273

sfranzis commented:
> Hi, I would like the "zero point" settable. For example:
> 
> * the current value is 3
> * The minimum value is -5
> * The maximum value is 5
> * the neutral point is manually set to -1
> 
> The gauge would then be drawn from -1 to 3
> 
> This could be useful to show how far away from a setpoint the current value is, e.g. a speed setting and the actual speed.

Hi, I understand your requirement. But this needs changes in the UI and model and I'm not yet deep enough in the complete code.

gguillotte-grafana reviewed (commented):
Docs additions were removed, so there's nothing for me to review.

sfranzis committed:
Merge branch 'grafana:main' into #38273

JoaoSilvaGrafana commented on public/vendor/flot/jquery.flot.gauge.js:
```diff
@@ -582,7 +637,7 @@
          * @param  {Object} textOptions the option of the text
          * @param  {Number} [a] the angle of the value drawn
          */
-        function drawText(x, y, id, text, textOptions, a) {
+         function drawText(x, y, id, text, textOptions, a) {
```
Space can be removed

tclausing replied:
Also check for positive max?
```
            var hasNegative = gaugeOptionsi.gauge.min < 0.0
            var hasPositive = gaugeOptionsi.gauge.max > 0.0
            if (hasNegative && hasPositive) {
```

sfranzis replied:
hasPositive implemented

JoaoSilvaGrafana reviewed (commented):
Good job on adding the 0 point! Just one small comment and could you merge with main as there is a bug on the NumberInput fixed if you merge.

Also I asked the team and the UX team has mentioned if we can add a "0" next to the line. I noticed that if we turn the "Show threshold labels" toggle on it already shows labels for min/max/thresholds, so I thought we could make use of that.

I think we can achieve that by adding
```
            var hasNegative = gaugeOptionsi.gauge.min < 0.0
            if (hasNegative) {
                drawThresholdValue(gaugeOptionsi, layout, cellLayout, "Zero" + i, 0, calculateAngle(gaugeOptionsi, layout, 0));
            }
```
to the start of the `drawThresholdValues` function, then it looks like this:
<img width="523" alt="Screenshot 2022-11-09 at 18 56 14" src="https://user-images.githubusercontent.com/100691367/200916989-2d6accb8-2bb5-4705-9b8a-5693b8c24942.png">

what do you think?

JoaoSilvaGrafana commented:
> Hi, I would like the "zero point" settable. For example:
> 
> * the current value is 3
> * The minimum value is -5
> * The maximum value is 5
> * the neutral point is manually set to -1
> 
> The gauge would then be drawn from -1 to 3
> 
> This could be useful to show how far away from a setpoint the current value is, e.g. a speed setting and the actual speed.

This is definitely more complicated, I think this can be a separate feature request because it would take longer to complete.

sfranzis committed:
zero option

sfranzis committed:
Revert "zero option"

This reverts commit b31bbdbf83ec18e31802867fd5102feb94f99be6.

sfranzis committed:
zero option

sfranzis committed:
Merge branch 'grafana:main' into #38273

sfranzis committed:
Merge branch 'grafana:main' into feature/zero-option

sfranzis committed:
threshold value for zero

sfranzis commented:
> Good job on adding the 0 point! Just one small comment and could you merge with main as there is a bug on the NumberInput fixed if you merge.
> 
> Also I asked the team and the UX team has mentioned if we can add a "0" next to the line. I noticed that if we turn the "Show threshold labels" toggle on it already shows labels for min/max/thresholds, so I thought we could make use of that.
> 
> I think we can achieve that by adding
> 
> ```
>             var hasNegative = gaugeOptionsi.gauge.min < 0.0
>             if (hasNegative) {
>                 drawThresholdValue(gaugeOptionsi, layout, cellLayout, "Zero" + i, 0, calculateAngle(gaugeOptionsi, layout, 0));
>             }
> ```
> 
> what do you think?

Hi Joao,

thanks for your review and the sync with UX team. I implemented your suggestions, synced my fork with main and pushed. 

Stefan

sfranzis commented:
> > Hi, I would like the "zero point" settable. For example:
> >  ....
> > This could be useful to show how far away from a setpoint the current value is, e.g. a speed setting and the actual speed.
> 
> This is definitely more complicated, I think this can be a separate feature request because it would take longer to complete.

I agree with this. However, in the meantime I already started working on it. ;-)
Still I would like to see my first implementation in a release and make a new request for the settable zero point.

For those who can't wait, see here: https://github.com/sfranzis/grafana/tree/feature/zero-option

Stefan

sfranzis committed:
fix merge

sfranzis committed:
remove space

sfranzis committed:
add newline at eof

sfranzis committed:
remove category

sfranzis committed:
remove space

sfranzis committed:
remove newline

sfranzis committed:
add newline

sfranzis committed:
draw neutral threshold value

sfranzis committed:
draw neutral threshold value

sfranzis committed:
Merge branch 'feature/zero-option' of https://github.com/sfranzis/grafana into feature/zero-option

JoaoSilvaGrafana commented:
Hey @sfranzis just to keep you up to date. We are discussing internally about adding an option here so that the default behavior is what it looked like before, in order to prevent adding breaking changes. Then if you turned that option on it would look like the changes in your PR.
Very sorry that the PR is taking so long to get to completion, we all think it's a really valuable change! Just want to make sure we merge this without any breaking changes.

sfranzis commented:
Hi @JoaoSilvaGrafana 
I have already an Implementation to set the neutral value. If you don't put in a value it has the default behavior as before. This offers also the ability to set neutral values other than 0.
<img width="1107" alt="image" src="https://user-images.githubusercontent.com/12957935/201911856-8f1ea993-00ba-46e9-9a12-88522d74835d.png">

If you like I can merge the code or open a second pull request.

Regards
Stefan

JoaoSilvaGrafana commented:
Oh fantastic!
Could you merge that with this branch then? We are having a quick call with UX tomorrow about this and I can demo it to them 😄 
Thanks and great job!

sfranzis commented:
Example use case for neutral value different from 0:
Monitoring the divergence of line frequency (50.0 Hz in Europe 60.0 Hz in the US and other countries)
<img width="1173" alt="image" src="https://user-images.githubusercontent.com/12957935/201913620-cc5d2951-b4f7-4604-8760-eb36fc60a722.png">

sfranzis commented:
> Oh fantastic!
> Could you merge that with this branch then? We are having a quick call with UX tomorrow about this and I can demo it to them 😄
> Thanks and great job!

Ok, I will do it later today when I have no disturbance. :-)

sfranzis committed:
Make neutral value configurable

sfranzis committed:
fixing merge glitch

sfranzis committed:
fixed type check

sfranzis committed:
Merge branch 'grafana:main' into #38273

sfranzis commented:
Hey @JoaoSilvaGrafana, 
merged & synced with main. Build passed. If you accept my changes I will also add a few lines to documentation.
Regards :-)

JoaoSilvaGrafana commented on packages/grafana-ui/src/components/Gauge/Gauge.tsx:
```diff
@@ -22,6 +22,7 @@ export interface Props {
   field: FieldConfig;
   showThresholdMarkers: boolean;
   showThresholdLabels: boolean;
+  neutral: Number;
```
Also optional as it can be `undefined`

JoaoSilvaGrafana commented on public/app/plugins/panel/gauge/models.cue:
```diff
@@ -29,6 +29,7 @@ Panel: thema.#Lineage & {
 						ui.SingleStatBaseOptions
 						showThresholdLabels:  bool | *false
 						showThresholdMarkers: bool | *true
+						neutral: number
```
Should be optional here as well

JoaoSilvaGrafana commented on public/vendor/flot/jquery.flot.gauge.js:
```diff
@@ -529,6 +584,11 @@
             }
+            
+            var neutral = gaugeOptionsi.gauge.neutralValue;
+            if (neutral != null) {
```
This needs a check for if neutral > min, to avoid things like this (notice the -10 and -11 overlapping on the left):

<img width="1275" alt="Screenshot 2022-11-16 at 16 49 11" src="https://user-images.githubusercontent.com/100691367/202242518-ce19c70f-9381-41a6-a8d4-fdfb4eb9abbe.png">

tclausing replied:
&& neutral < max

JoaoSilvaGrafana reviewed (commented):
Great job on this! I showed your work to some colleagues and they really like the NumberInput option you created! 🎉  Just left some comments on it

sfranzis committed:
implement suggestions from review

sfranzis committed:
implement suggestions from review

sfranzis commented:
Thanks for your review and comments. Implemented it and pushed.
Just one idea left. I would like to have the Neutral Input also in the Overrides section. There you find for some panels only Standard options. Do you have an idea how to implement this. I've already searched a few hours. 🙈
With this you could build Gauge panels with different data (Voltage neutral at 230, Frequency at 50):
<img width="960" alt="image" src="https://user-images.githubusercontent.com/12957935/202266388-7079dadd-8d46-464b-8057-ac923aaf19c0.png">

sfranzis committed:
run make gen-cue

JoaoSilvaGrafana commented:
> Do you have an idea or hint how to implement this?

I'm not really sure how to implement overrides no... I had a quick look around and couldn't get it to work properly. I think what we have right now in this PR is already really valuable for us to merge it, and then if you figure out the overrides you can do it as a separate PR what do you think? 😄

sfranzis commented:
> 
> I think what we have right now in this PR is already really valuable for us to merge it, and then if you figure out the overrides you can do it as a separate PR what do you think? 😄 

Go for it 👍 I'm looking forward to see the current change in an official release. 😊

sfranzis committed:
move neutral to custom field config

sfranzis commented:
@JoaoSilvaGrafana found the override thing.
You have to implement the property as a custom field config, not as an option.
Changed it and pushed. If I implemented in a later request, it would get a breaking change.
Please see again and tell me what you think. 🙂

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/12957935/202637025-a9c7d529-4588-448d-b05f-d689707cd298.png">

JoaoSilvaGrafana reviewed (approved):
Great job on this, we really appreciate all the work you've put into it 👏
